### DISCUSSION TITLE
NtQueryInformationFile(..., FileCaseSensitiveInformation)

### DISCUSSION THREAD

**Discussion by DenisAnisimov** [dokan-dev/dokany#1179] [General]
I'm trying to determine the value of FileCaseSensitiveInformation with code like this:

```
function DriveCaseSensitive(ADrive: Char): Boolean;
var
  DriveHandle: THandle;
  IOStatus: TIOStatusBlock;
  Flags: DWORD;
  ErrorCode: NTSTATUS;
begin
  Result := False;
  DriveHandle := CreateFile(PChar(ADrive + ':\'), FILE_READ_ATTRIBUTES,
    FILE_SHARE_READ or FILE_SHARE_WRITE, nil, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, 0);
  if DriveHandle <> INVALID_HANDLE_VALUE then
    begin
      ErrorCode := NtQueryInformationFile(DriveHandle, @IOStatus, @Flags, SizeOf(Flags), FileCaseSensitiveInformation); // Have 0xC000000D here
      if ErrorCode = STATUS_SUCCESS then
        Result := Flags and FILE_CS_FLAG_CASE_SENSITIVE_DIR <> 0;
      CloseHandle(DriveHandle);
    end;
end;
```
It works correctly with NTFS, but does not work with Docany based drives.

I also want to ask, is there a way to determine using WinApi or NativeAPI that the drive is Docany based?

**Comment by Olof-Lagerkvist** [comment]
> I'm trying to determine the value of FileCaseSensitiveInformation with code like this:
> 
> It works correctly with NTFS, but does not work with Docany based drives.

Use `GetVolumeInformation` instead and check `lpFileSystemFlags` for flags like `FILE_CASE_SENSITIVE_SEARCH` and `FILE_CASE_PRESERVED_NAMES`.

> I also want to ask, is there a way to determine using WinApi or NativeAPI that the drive is Docany based?

Not that easily. But if it is your own Dokan drive that you want to detect somewhere, you can set a file system name from Dokan that can later be identified in file system name parameter from `GetVolumeInformation` calls.

**Reply by DenisAnisimov** [reply]
> Use GetVolumeInformation instead and check lpFileSystemFlags for flags like FILE_CASE_SENSITIVE_SEARCH and FILE_CASE_PRESERVED_NAMES.

NTFS returns FILE_CASE_SENSITIVE_SEARCH, but returns FileCaseSensitiveInformation as False. As far as I understand, FileCaseSensitiveInformation is not the same as FILE_CASE_SENSITIVE_SEARCH. FILE_CASE_SENSITIVE_SEARCH is the RAW level, the FS says it CAN use CaseSensitive names. And FileCaseSensitiveInformation is the logical level, the FS says that it DOES use CaseSensitive names.

Or do you mean that I should use GetVolumeInformation in case NtQueryInformationFile returns an error?

> But if it is your own Dokan drive...

I'm developing my FS based on Docany. But at the same time, I am developing a utility for testing FS, which compares various aspects of the behavior of my FS (error codes, working with file dates, etc.) and the behavior of NTFS. In some cases, Docany returns error codes other than the NTFS error codes, and my utility displays this. The utility has an option "Ignore Docany specifics/bugs" that can be enabled before the test, but I would like the utility to automatically detect that the drive is Docany-based.

**Comment by Olof-Lagerkvist** [comment]
> > Use GetVolumeInformation instead and check lpFileSystemFlags for flags like FILE_CASE_SENSITIVE_SEARCH and FILE_CASE_PRESERVED_NAMES.
> 
> NTFS returns FILE_CASE_SENSITIVE_SEARCH, but returns FileCaseSensitiveInformation as False. As far as I understand, FileCaseSensitiveInformation is not the same as FILE_CASE_SENSITIVE_SEARCH. FILE_CASE_SENSITIVE_SEARCH is the RAW level, the FS says it CAN use CaseSensitive names. And FileCaseSensitiveInformation is the logical level, the FS says that it DOES use CaseSensitive names.

`FILE_CASE_SENSITIVE_SEARCH` indicates that the file system will distinguish between upper- and lower-case letters when searching for a file. `FILE_CASE_PRESERVED_NAMES` indicates that case is preserved in file names. For latest NTFS versions, `FILE_CASE_SENSITIVE_SEARCH` is set in the flags to indicate that the volume can support this. However, to actually use this feature you would need to activate it for the particular subdirectory using `fsutil file setCaseSensitiveInfo` etc. It would theoretically be possible to implement something like this for Dokan drives as well, but currently the only supported indication is volume-wide support or not support for searching with case sensitivity.

> Or do you mean that I should use GetVolumeInformation in case NtQueryInformationFile returns an error?

I mean that you should not assume that all query classes for `NtQueryInformationFile` or `NtQueryVolumeInformationFile` are implemented by a particular file system. However, `GetVolumeInformation` should always be implemented in a compatible way.

> > But if it is your own Dokan drive...
> 
> I'm developing my FS based on Docany. But at the same time, I am developing a utility for testing FS, which compares various aspects of the behavior of my FS (error codes, working with file dates, etc.) and the behavior of NTFS. In some cases, Docany returns error codes other than the NTFS error codes, and my utility displays this. The utility has an option "Ignore Docany specifics/bugs" that can be enabled before the test, but I would like the utility to automatically detect that the drive is Docany-based.

Easiest way then is probably to just set your own file system name from Dokan side and then look for that file system name in calls to `GetVolumeInformation` in your test application.

**Reply by DenisAnisimov** [reply]
>Easiest way then is probably to just set your own file system name from Dokan side


Later I plan to make the utility publicly available, so I would prefer a universal automatic solution rather than manually setting the name. But anyway, thanks for the answers.
